Source/Core/Helpers.cs

```diff
@@ -264,31 +265,19 @@ public static void ExtraTraceInformation(CoreOptions options, string point)
       }
     }
 
-    private static readonly HashSet<string> UsedLogNames = new();
+    private static readonly ConcurrentDictionary<string, int> UsedLogNames = new();
     public static (string fileName, bool reused) GetLogFilename(string descriptiveName, string filename, bool allowReuse)
     {
-      filename = SubstituteAtProc(descriptiveName, cce.NonNull(filename));
-      var curFilename = filename;
-      var reused = false;
+      filename = SubstituteAtPROC(descriptiveName, cce.NonNull(filename));
 
-      lock (UsedLogNames) {
-        int n = 1;
-        while (!allowReuse && UsedLogNames.Contains(curFilename)) {
-          curFilename = filename + "." + n++;
-        }
-
-        if (UsedLogNames.Contains(curFilename)) {
-          reused = true;
-        }
-
-        UsedLogNames.Add(curFilename);
-      }
+      var reused = allowReuse && UsedLogNames.ContainsKey(filename);
+      var index = UsedLogNames.AddOrUpdate(filename, 0, (_, i) => allowReuse ? i : i + 1);
```

Comment:
This code is thread safe because the race between `ContainsKey` and `AddOrUpdate` only happens in the nonconcurrent case (`allowReuse`), but maybe this will be confusing to readers?  I *think* `reused` is the same as `allowReuse && index > 0`?

Reply:
Ah, no, sorry, I misread the code.

Reply:
I guess you could leave things as is, or maybe use this:

```
      var index = UsedLogNames.AddOrUpdate(filename, 0, (_, i) => allowReuse ? -1 : i + 1);
      var reused = index == -1
```

Or alternatively use this:

```
if (allowReuse) {
  return filename, UsedLogNames.ContainsKey(filename);
}

… rest of the code with reused = false
```

I'm assuming you can't mix calls with and without allowReuse in a single process?

Reply:
New version is perfect :)